src/osal/linux/osal_eth.c

```diff
@@ -27,109 +27,112 @@
 #include <sys/ioctl.h>
 #include <netpacket/packet.h>
 
-static int os_eth_recv(uint16_t type, os_buf_t * p)
-{
-   int handled = 0;
-
-   switch (type)
-   {
-   case OS_ETHTYPE_PROFINET:
-   case OS_ETHTYPE_LLDP:
-      handled = pf_eth_recv(p, p->len);
-      break;
-   case OS_ETHTYPE_ETHERCAT:
-      handled = 0;
-      break;
-
-   default:
-      break;
-   }
-
-   return handled;
-}
 
-static void os_eth_task(void * arg)
+/**
+ * @internal
+ * Run a thread that listens to incoming raw Ethernet sockets.
+ * Delegate the actual work to thread_arg->callback
+ *
+ * This is a function to be passed into os_thread_create()
+ * Do not change the argument types.
+ *
+ * @param thread_arg     InOut: Will be converted to os_eth_handle_t
+ */
+static void os_eth_task(
+   void *                  thread_arg)
 {
-   int * sock = (int *) arg;
-   ssize_t readlen;
+   os_eth_handle_t         *eth_handle = thread_arg;
+   ssize_t                 readlen;
+   int                     handled = 0;
 
-   os_buf_t * p = os_buf_alloc (1522);
+   os_buf_t *p = os_buf_alloc(OS_BUF_MAX_SIZE);
    assert(p != NULL);
 
    while (1)
    {
-      readlen = recv (*sock, p->payload, p->len, 0);
+      readlen = recv(eth_handle->socket, p->payload, p->len, 0);
       if(readlen == -1)
          continue;
       p->len = readlen;
-      if (os_eth_recv (OS_ETHTYPE_PROFINET, p) == 1)
+
+      if (eth_handle->callback != NULL)
+      {
+         handled = eth_handle->callback(eth_handle->arg, p);
+      }
+      else
+      {
+         handled = 0;
+      }
+
+      if (handled == 1)
       {
-         fflush(stdout);
-         p = os_buf_alloc (1522);
+         p = os_buf_alloc(OS_BUF_MAX_SIZE);
          assert(p != NULL);
       }
    }
 }
 
-int os_eth_init(const char * if_name)
+os_eth_handle_t* os_eth_init(
+   const char              *if_name,
+   os_eth_callback_t       *callback,
+   void                    *arg)
 {
-
-   int * sock;
-   int i;
-   struct ifreq ifr;
-   struct sockaddr_ll sll;
-   int ifindex;
-   struct timeval timeout;
-
-   sock = calloc (1, sizeof(int));
-   *sock = -1;
-
-   *sock = socket(PF_PACKET, SOCK_RAW, htons(OS_ETHTYPE_PROFINET));
-
-   timeout.tv_sec =  0;
+   os_eth_handle_t         *handle;
+   int                     i;
+   struct ifreq            ifr;
+   struct sockaddr_ll      sll;
+   int                     ifindex;
+   struct timeval          timeout;
+
+   handle = malloc(sizeof(os_eth_handle_t));
```

Comment:
Check that malloc does not return NULL (check all calls of malloc in this commit).

Reply:
Done